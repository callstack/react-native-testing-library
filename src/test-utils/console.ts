import { format } from 'util';

export function excludeConsoleMessage(logFn: (...args: unknown[]) => void, excludeMessage: string) {
  return (...args: unknown[]) => {
    const message = format(...args);
    if (message.includes(excludeMessage)) {
      return;
    }

    logFn(...args);
  };
}
