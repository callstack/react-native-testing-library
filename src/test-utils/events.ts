interface EventEntry {
  name: string;
  payload: any;
}

export function createEventLogger() {
  const events: EventEntry[] = [];
  const logEvent = (name: string) => {
    return (event: unknown) => {
      const eventEntry: EventEntry = {
        name,
        payload: event,
      };

      events.push(eventEntry);
    };
  };

  return { events, logEvent };
}
