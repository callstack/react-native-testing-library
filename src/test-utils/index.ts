export * from './console';
export * from './events';
