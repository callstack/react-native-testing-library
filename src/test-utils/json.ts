import type { ReactTestRendererJSON } from 'react-test-renderer';

type JsonPropsMapper = {
  [key: string]: unknown;
};

export function mapJsonProps<T extends ReactTestRendererJSON | ReactTestRendererJSON[] | null>(
  element: T,
  mapper: JsonPropsMapper,
): T {
  if (Array.isArray(element)) {
    return element.map((e) => mapJsonProps(e, mapper)) as T;
  }

  if (!element) {
    return element;
  }

  const resultProps = { ...element.props };
  Object.keys(mapper).forEach((key) => {
    if (key in element.props) {
      resultProps[key] = mapper[key];
    }
  });

  const resultElement = { ...element, props: resultProps };
  Object.defineProperty(resultElement, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return resultElement;
}
