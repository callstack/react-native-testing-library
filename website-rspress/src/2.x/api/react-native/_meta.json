[
  {
    "type": "file",
    "name": "index",
    "label": "Table of Contents"
  },
  {
    "type": "dir",
    "name": "classes",
    "label": "Classes"
  },
  {
    "type": "dir",
    "name": "interfaces",
    "label": "Interfaces"
  },
  {
    "type": "dir",
    "name": "types",
    "label": "Types"
  }
]
