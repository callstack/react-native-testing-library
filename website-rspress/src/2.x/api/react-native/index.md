# Re.Pack React Native APIs

## Table of contents

### Classes

- [Chunk](./classes/Chunk.md)
- [ChunkManager](./classes/ChunkManager.md)

### Interfaces

- [ChunkConfig](./interfaces/ChunkConfig.md)
- [ChunkManagerConfig](./interfaces/ChunkManagerConfig.md)
- [RemoteChunkLocation](./interfaces/RemoteChunkLocation.md)
- [StorageApi](./interfaces/StorageApi.md)

### Type aliases

- [RemoteChunkResolver](./types/RemoteChunkResolver.md)
