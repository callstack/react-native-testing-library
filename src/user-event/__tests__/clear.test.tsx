import * as React from 'react';
import { TextInput, TextInputProps, View } from 'react-native';
import { createEventLogger } from '../../test-utils';
import { render, userEvent, screen } from '../..';

beforeEach(() => {
  jest.useRealTimers();
});

function renderTextInputWithToolkit(props: TextInputProps = {}) {
  const { events, logEvent } = createEventLogger();

  render(
    <TextInput
      testID="input"
      onFocus={logEvent('focus')}
      onBlur={logEvent('blur')}
      onPressIn={logEvent('pressIn')}
      onPressOut={logEvent('pressOut')}
      onChange={logEvent('change')}
      onChangeText={logEvent('changeText')}
      onKeyPress={logEvent('keyPress')}
      onTextInput={logEvent('textInput')}
      onSelectionChange={logEvent('selectionChange')}
      onSubmitEditing={logEvent('submitEditing')}
      onEndEditing={logEvent('endEditing')}
      onContentSizeChange={logEvent('contentSizeChange')}
      {...props}
    />,
  );

  const textInput = screen.getByTestId('input');

  return {
    events,
    textInput,
  };
}

describe('clear()', () => {
  it('supports basic case', async () => {
    jest.spyOn(Date, 'now').mockImplementation(() => 100100100100);
    const { textInput, events } = renderTextInputWithToolkit({
      value: 'Hello!',
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'focus',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('value: "Hello!');
  });

  it.each(['modern', 'legacy'])('works with %s fake timers', async (type) => {
    jest.useFakeTimers({ legacyFakeTimers: type === 'legacy' });
    const { textInput, events } = renderTextInputWithToolkit({
      value: 'Hello!',
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'focus',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);
  });

  it('supports defaultValue prop', async () => {
    const { textInput, events } = renderTextInputWithToolkit({
      defaultValue: 'Hello Default!',
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'focus',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('defaultValue: "Hello Default!"');
  });

  it('does respect editable prop', async () => {
    const { textInput } = renderTextInputWithToolkit({
      value: 'Hello!',
      editable: false,
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    expect(textInput.props.value).toBe('Hello!');
  });

  it('does respect pointer-events prop', async () => {
    const { textInput } = renderTextInputWithToolkit({
      value: 'Hello!',
      pointerEvents: 'none',
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    expect(textInput.props.value).toBe('Hello!');
  });

  it('supports multiline', async () => {
    const { textInput, events } = renderTextInputWithToolkit({
      value: 'Hello World!\nHow are you?',
      multiline: true,
    });

    const user = userEvent.setup();
    await user.clear(textInput);

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'focus',
      'selectionChange',
      'keyPress',
      'textInput',
      'change',
      'changeText',
      'selectionChange',
      'contentSizeChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('value: "Hello World!\nHow are you?" multiline: true,');
  });

  it('works when not all events have handlers', async () => {
    const { events, logEvent } = createEventLogger();
    render(
      <TextInput
        testID="input"
        onChangeText={logEvent('changeText')}
        onEndEditing={logEvent('endEditing')}
      />,
    );

    const user = userEvent.setup();
    await user.clear(screen.getByTestId('input'));

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual(['changeText', 'endEditing']);

    expect(events).toMatchSnapshot();
  });

  it('does NOT work on View', async () => {
    render(<View testID="input" />);

    const user = userEvent.setup();
    await expect(
      user.clear(screen.getByTestId('input')),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"clear() only supports host "TextInput" elements. Passed element has type: "View"."`,
    );
  });

  // View that ignores props type checking
  const AnyView = View as React.ComponentType<any>;

  it('does NOT bubble up', async () => {
    const parentHandler = jest.fn();
    render(
      <AnyView
        onChangeText={parentHandler}
        onChange={parentHandler}
        onKeyPress={parentHandler}
        onTextInput={parentHandler}
        onFocus={parentHandler}
        onBlur={parentHandler}
        onEndEditing={parentHandler}
        onPressIn={parentHandler}
        onPressOut={parentHandler}
      >
        <TextInput testID="input" />
      </AnyView>,
    );

    const user = userEvent.setup();
    await user.clear(screen.getByTestId('input'));
    expect(parentHandler).not.toHaveBeenCalled();
  });
});
