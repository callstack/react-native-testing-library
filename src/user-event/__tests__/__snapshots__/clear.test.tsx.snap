// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clear() supports basic case: value: "Hello! 1`] = `
[
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 6,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "key": "Backspace",
      },
      "target": {},
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "changeText",
    "payload": "",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 0,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
]
`;

exports[`clear() supports defaultValue prop: defaultValue: "Hello Default!" 1`] = `
[
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 14,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "key": "Backspace",
      },
      "target": {},
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "changeText",
    "payload": "",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 0,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
]
`;

exports[`clear() supports multiline: value: "Hello World!
How are you?" multiline: true, 1`] = `
[
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 25,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "key": "Backspace",
      },
      "target": {},
    },
  },
  {
    "name": "textInput",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "previousText": "Hello World!
How are you?",
        "range": {
          "end": 0,
          "start": 0,
        },
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "changeText",
    "payload": "",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "selection": {
          "end": 0,
          "start": 0,
        },
      },
      "target": {},
    },
  },
  {
    "name": "contentSizeChange",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "contentSize": {
          "height": 16,
          "width": 0,
        },
        "target": 0,
      },
      "target": {},
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
      },
      "target": {},
    },
  },
]
`;

exports[`clear() works when not all events have handlers 1`] = `
[
  {
    "name": "changeText",
    "payload": "",
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "nativeEvent": {
        "target": 0,
        "text": "",
      },
      "target": {},
    },
  },
]
`;
